enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,397 @@
+  - https://github.com/openshift/enhancements/pull/384
+replaces: []
+superseded-by: []
+---
+
+Start by filling out this header template with metadata for this enhancement.
+
+- `reviewers`: This can be anyone that has an interest in this work.
+
+- `approvers`: All enhancements must be approved, but the appropriate people to approve a given
+  enhancement depends on its scope. If an enhancement is limited in scope to a given team or
+  component, then a peer or lead on that team or pillar is an appropriate approver. If an
+  enhancement captures something more broad in scope, then a member of the OpenShift architects team
+  or someone they delegate would be appropriate. Examples would be something that changes the
+  definition of OpenShift in some way, adds a new required dependency, or changes the way customers
+  are supported. Use your best judgement to determine the level of approval needed. If you’re not
+  sure, just leave it blank and ask for input during review.
+
+# Subscription Injection Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+      [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will allow any workload to access RHEL subscription content via a pod annotation.
+
+## Motivation
+
+Containers can download RHEL content if their hosts have subscriptions attached. In 3.x, all nodes
+were assumed to be RHEL 7 nodes that were subscribed individually. The information needed to access
+subscription content (entitlement keys and `redhat.repo` configurations) was symbolically linked
+into the running container at a well-known location that yum/dnf could find. This capability existed
+in our patch of Docker, and was carried over into Red Hat’s container toolchain used on OpenShift 4
+(cri-o, podman, buildah, etc.).
+
+In OpenShift 4 the preferred operating system (RHCOS) is not capable of attaching subscriptions
+individually. To access subscription content, containers must be provided entitlement keys and
+`redhat.repo` configurations via other means. Enabling this capability needs to be simple and
+straightforward.
+
+### Goals
+
+1. Inject entitlement keys and `redhat.repo` configurations into a pod spec.
+2. Define APIs and conventions that allow cluster admins to add the required information to a
+   cluster (entitlement keys and `redhat.repo` configurations).
+
+### Non-Goals
+
+1. Automatically install entitlement keys and `redhat.repo` configurations onto a subscribed
+   cluster.
+
+## Proposal
+
+### User Stories
+
+#### Add debug tools to containers
+
+As a developer debugging my application I want access to RHEL subscription content in my containers
+So that I can install debugging tools from Red Hat via `yum install`
```

Comment:
Subscribing the host solves this.

Reply:
my understand is there is no such thing as subscribing the host for rhcos nodes.

Reply:
No reason not to. Just hasn't been "made a thing" and that's what this operator should do unless there's an easy way without an operator at all.

Reply:
(I even subscribe my Fedora laptop so I can run RHEL containers.)

Reply:
> No reason not to. 

the reason not to is administrative overhead of managing subscriptions on those machines (and assigning subscriptions to those machines).  I'm not sufficiently versed in the business side of things, but i think it's been a deliberate choice not to make admins manage per machine subscriptions for openshift clusters.

i'll let @derekwaynecarr or @eparis weigh in here, but i think that approach was discarded pretty early on in the "how do we enable entitlements for pods/builds in OCP4" pretty early on.

the goal here is to have a single subscription for the cluster which all workloads can leverage as needed.

Reply:
Then just distribute that single credential across the cluster in the well-known locations on the hosts so  that cri-o will pick it up. That's even easier!

Reply:
doing this at the crio level means end users have no choice about whether or not content is injected into their pods and admins have no control over who is allowed to access those creds (anyone who can launch a pod can do it).

among other things, it makes any sort of hosted multitenant cluster untenable.

Reply:
That's correct, the idea of subscribing individual hosts has been rejected. Host subscriptions bring no value to the host. Hosts are quite ephemeral in the changing IT landscape and managing such subscriptions does not make sense.

Reply:
Note too that in #384 I have added multi-cluster administration as a user story. Distributing entitlement credentials needs to scale to potentially thousands of clusters, each with as few as 1 node or as many as 10,000.